File: ros2_self_msg_srv/src/my_interfaces_py/my_interfaces_py/service_server.py
#!/usr/bin/env python3
"""服务端示例 - Python版本：使用自定义srv"""
import rclpy
from rclpy.node import Node
from my_interfaces.srv import AddThreeInts


class ServiceServer(Node):
    """三个整数相加服务端"""

    def __init__(self):
        super().__init__('service_server')
        # 创建服务，使用自定义服务类型
        self.srv = self.create_service(
            AddThreeInts,
            'add_three_ints',
            self.add_callback
        )
        self.get_logger().info('服务端已启动，等待请求...')

    def add_callback(self, request, response):
        """服务回调：计算三个整数的和"""
        response.sum = request.a + request.b + request.c
        self.get_logger().info(
            f'收到请求: a={request.a}, b={request.b}, c={request.c}'
        )
        self.get_logger().info(f'返回结果: sum={response.sum}')
        return response


def main(args=None):
    rclpy.init(args=args)
    server = ServiceServer()
    try:
        rclpy.spin(server)
    except KeyboardInterrupt:
        pass
    finally:
        server.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: ros2_self_msg_srv/src/my_interfaces_cpp/src/service_server.cpp
/**
 * 服务端示例 - C++版本：使用自定义srv
 */
#include "rclcpp/rclcpp.hpp"
#include "my_interfaces/srv/add_three_ints.hpp"

#include <memory>

class ServiceServer : public rclcpp::Node
{
public:
    ServiceServer() : Node("service_server")
    {
        // 创建服务，使用自定义服务类型
        srv_ = this->create_service<my_interfaces::srv::AddThreeInts>(
            "add_three_ints",
            std::bind(&ServiceServer::add, this, std::placeholders::_1, std::placeholders::_2));

        RCLCPP_INFO(this->get_logger(), "服务端已启动，等待请求...");
    }

private:
    void add(
        const std::shared_ptr<my_interfaces::srv::AddThreeInts::Request> request,
        std::shared_ptr<my_interfaces::srv::AddThreeInts::Response> response)
    {
        response->sum = request->a + request->b + request->c;
        RCLCPP_INFO(this->get_logger(), 
            "收到请求: a=%ld, b=%ld, c=%ld", 
            request->a, request->b, request->c);
        RCLCPP_INFO(this->get_logger(), "返回结果: sum=%ld", response->sum);
    }

    rclcpp::Service<my_interfaces::srv::AddThreeInts>::SharedPtr srv_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<ServiceServer>());
    rclcpp::shutdown();
    return 0;
}
